File: estilofit-web/src/services/categoryService.ts
import { api } from "../lib/api";
import type { Category, CategoryRequest } from "../types/category";

export const categoryService = {
  async list(onlyActive: boolean): Promise<Category[]> {
    const { data } = await api.get<Category[]>("/categories", {
      params: { onlyActive },
    });
    return data;
  },

  async create(request: CategoryRequest): Promise<Category> {
    const { data } = await api.post<Category>("/categories", request);
    return data;
  },

  async rename(id: string, request: CategoryRequest): Promise<Category> {
    const { data } = await api.put<Category>(`/categories/${id}`, request);
    return data;
  },

  async updateStatus(id: string, active: boolean): Promise<Category> {
    const { data } = await api.patch<Category>(`/categories/${id}/status`, { active });
    return data;
  },
};

File: estilofit-api/src/main/kotlin/br/com/estilofitudi/category/service/CategoryService.kt
package br.com.estilofitudi.category.service

import br.com.estilofitudi.category.domain.Category
import br.com.estilofitudi.category.dto.CategoryRequest
import br.com.estilofitudi.category.dto.CategoryResponse
import br.com.estilofitudi.category.dto.toResponse
import br.com.estilofitudi.category.repository.CategoryRepository
import br.com.estilofitudi.shared.exception.DataConflictException
import br.com.estilofitudi.shared.exception.EntityNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true)
class CategoryService(
    private val categoryRepository: CategoryRepository,
) {

    fun findAll(onlyActive: Boolean): List<CategoryResponse> {
        val categories = if (onlyActive) {
            categoryRepository.findAllByActiveTrueOrderByNameAsc()
        } else {
            categoryRepository.findAllByOrderByNameAsc()
        }
        return categories.map { it.toResponse() }
    }

    @Transactional
    fun create(request: CategoryRequest): CategoryResponse {
        val name = request.name.trim()
        if (categoryRepository.existsByNameIgnoreCase(name)) {
            throw DataConflictException("Categoria '$name' já existe")
        }
        val category = Category(name = name)
        return categoryRepository.save(category).toResponse()
    }

    @Transactional
    fun rename(id: UUID, request: CategoryRequest): CategoryResponse {
        val category = categoryRepository.findById(id)
            .orElseThrow { EntityNotFoundException("Categoria", id) }

        val name = request.name.trim()
        if (categoryRepository.existsByNameIgnoreCaseAndIdNot(name, id)) {
            throw DataConflictException("Categoria '$name' já existe")
        }

        category.name = name
        return categoryRepository.save(category).toResponse()
    }

    @Transactional
    fun updateStatus(id: UUID, active: Boolean): CategoryResponse {
        val category = categoryRepository.findById(id)
            .orElseThrow { EntityNotFoundException("Categoria", id) }
        category.active = active
        return categoryRepository.save(category).toResponse()
    }
}
